Fix the bugs of float/double precision error in the assertions

YeeTone opened the pull request:
#### Check List:
* Fixes #2537
* Unit tests : YES
* Javadoc with a code example (on API only) : YES
* PR meets the [contributing guidelines](https://github.com/assertj/assertj-core/blob/main/CONTRIBUTING.md)

Considering the [issue#2537](https://github.com/assertj/assertj-core/issues/2537), the precision error can be fixed by using the combination of `String.valueOf` and `BigDecimal`.

Following the [contributing guidelines](https://github.com/assertj/assertj-core/blob/main/CONTRIBUTING.md) will make it easier for us to review and accept your PR.

This pull request fixes the bug that offset with float[] assertion doesn't have the same behavior as offset with float assertion, which is the main topic of [issue#2414](https://github.com/assertj/assertj-core/issues/2414)

joel-costigliola commented on src/main/java/org/assertj/core/api/ComparatorFactory.java:
```diff
@@ -12,8 +12,7 @@
+import java.math.BigDecimal;
 import java.util.Comparator;
 
 public class ComparatorFactory {
```
Can you create two test classes `ComparatorFactory_doubleComparatorWithPrecision_Test` and `ComparatorFactory_floatComparatorWithPrecision_Test`, each should have a parameterized test verifying the behavior of the comparator on different values. You can look at `Offset_Test` for parameterized test examples.

YeeTone replied:
OK. Within this week.

YeeTone replied:
> Can you create two test classes `ComparatorFactory_doubleComparatorWithPrecision_Test` and `ComparatorFactory_floatComparatorWithPrecision_Test`, each should have a parameterized test verifying the behavior of the comparator on different values. You can look at `Offset_Test` for parameterized test examples.

Can I design a super class named `ComparatorFactory_ComparatorWithPrecision_Test` to avoid the code duplication on these 2 testcases?

joel-costigliola commented on src/main/java/org/assertj/core/internal/Doubles.java:
```diff
@@ -62,7 +64,7 @@ protected Double NaN() {
   @Override
   protected Double absDiff(Double actual, Double other) {
-    return abs(other - actual);
+    return abs(new BigDecimal(String.valueOf(actual)).subtract(new BigDecimal(String.valueOf(other))).doubleValue());
```
Add some tests in `Doubles_assertIsCloseTo_Test` showing that the offset is properly honored.
Same thing for `Floats_assertIsCloseTo_Test`

joel-costigliola reviewed (changes requested):
Thanks @YeeTone for the PR!
There is some potential refactoring to do and more importantly tests to add to show the new fixed behavior ;-)

YeeTone commented:
OK. I have understood your idea. Please give me some time travel provide the testcases and refactoring!

YeeTone commented on src/test/java/org/example/test/AssertFloatArrayWithInRange_Test.java:
```diff
@@ -0,0 +1,28 @@
+
+import static org.assertj.core.api.Assertions.*;
+
+public class AssertFloatArrayWithInRange_Test {
```
Thanks, this test is not completed. I'll fix this soon.

YeeTone commented on src/main/java/org/assertj/core/api/ComparatorFactory.java:
```diff
@@ -25,7 +24,13 @@
+        BigDecimal abs = bd1.subtract(bd2).abs();
+        if (abs.compareTo(new BigDecimal(String.valueOf(precision))) <= 0) {
+          return 0;
+        }
```
Nice idea. I only focus on implementing the function but not on the code duplication. I will work on it! thanks.

YeeTone commented:
Sorry, I found some bugs yesterday and did not push the bug-fixed version.
The previous problem is that the combination of BigDecimal and String.valueOf cannot handle some special cases such as Nan or Infinity. I have already fixed the bug and pushed the newer version.

YeeTone commented:
I found I did not add the licence header information. I have added it. Please trigger the test again, thx.

YeeTone commented:
Could you run the test again? I found I did not delete an unnecessary testing class.

joel-costigliola commented:
@YeeTone don't worry about pitest, it's a bug in pitest

YeeTone commented:
> @YeeTone don't worry about pitest, it's a bug in pitest

Could you give me suggestions about my PR? Or considering merge it?

YeeTone commented:
> @YeeTone don't worry about pitest, it's a bug in pitest

Could you give me suggestions about my PR? Or considering merge it?@joel-costigliola

joel-costigliola commented:
I'll review it this week (or at least try to!)

YeeTone commented:
> I'll review it this week (or at least try to!)

Thank you very much!

joel-costigliola commented on src/main/java/org/assertj/core/internal/Numbers.java:
```diff
@@ -51,8 +49,8 @@ public Numbers(ComparisonStrategy comparisonStrategy) {
-   * 
-   * @param info contains information about the assertion.
+   *
+   * @param info   contains information about the assertion.
```
please revert al the unwanted formatting, this makes this PR hard to review because 90% of the changes in this file are irrelevant.
https://github.com/assertj/assertj-core/blob/main/CONTRIBUTING.md indicates how to get the proper formatting rules.

YeeTone replied:
I have fixed the formatting problem. Thx!

joel-costigliola commented on src/main/java/org/assertj/core/internal/Numbers.java:
```diff
@@ -261,6 +259,15 @@ public void assertIsNotCloseToPercentage(final AssertionInfo info, final NUMBER
+      return ((Double) value).isInfinite() || ((Double) value).isNaN();
+    else
+      return false;
+  }
```
this should go in `RealNumbers` and then you can use it `ComparatroFactory` (please avoid code duplication)

joel-costigliola replied:
why is this comment not addressed @YeeTone ?

joel-costigliola commented on src/test/java/org/assertj/core/internal/doubles/Doubles_assertIsCloseTo_Test.java:
```diff
@@ -73,8 +73,36 @@ void should_pass_if_actual_and_expected_are_NaN() {
+  void should_pass_if_actual_and_expected_are_equal_within_range2(){
+    doubles.assertIsCloseTo(someInfo(), 0.375d, 0.125, within(0.25));
+    doubles.assertIsCloseTo(someInfo(), 0.125d, 0.375, byLessThan(0.2500001));
+  }
```
these tests assertions belong either in `should_pass_if_difference_is_less_than_given_offset` or `should_pass_if_difference_is_equal_to_given_offset`, move the assertions in the proper test methods and remove these tests.

joel-costigliola commented on src/test/java/org/assertj/core/internal/floats/Floats_assertIsCloseTo_Test.java:
```diff
@@ -73,8 +73,37 @@ void should_pass_if_actual_and_expected_are_NaN() {
+  void should_pass_if_actual_and_expected_are_equal_within_range2(){
+    floats.assertIsCloseTo(someInfo(), 0.375f, 0.125f, within(0.25f));
+    floats.assertIsCloseTo(someInfo(), 0.125f, 0.375f, byLessThan(0.2500001f));
+  }
```
same comment as in the doubles tests

joel-costigliola commented on src/test/java/org/assertj/core/api/ComparatorFactory_floatComparatorWithPrecision.java:
```diff
@@ -0,0 +1,84 @@
+    assertThat(1).isEqualTo(comparator0_1_6f.compare(1.0000101f, 1f));
+    assertThat(-1).isEqualTo(comparator0_1_6f.compare(0.9999899f, 1f));
+    //assertThat(1).isEqualTo(comparator0_1_6f.compare(0.1111111111211, 0.111111111111));
+    //assertThat(-1).isEqualTo(comparator0_1_6f.compare(0.111111111111, 0.1111111111211));
```
don't leave comments like this, either remove them or if relevant uncomment

YeeTone replied:
Comments have been removed.

joel-costigliola commented on src/test/java/org/assertj/core/api/ComparatorFactory_floatComparatorWithPrecision.java:
```diff
@@ -0,0 +1,84 @@
+    assertThat(0).isEqualTo(comparator0_01f.compare(0.111f, 0.121f));
+    assertThat(0).isEqualTo(comparator0_01f.compare(0.121f, 0.111f));
+    assertThat(1).isEqualTo(comparator0_01f.compare(0.1211f, 0.111f));
+    assertThat(-1).isEqualTo(comparator0_01f.compare(0.111f, 0.1211f));
```
you are inverting actual and expected here, `comparator0_01f.compare(0.111f, 0.1211f)` is the object under test (actual) whereas `-1` is the expected value., this is why writing test with GIVEN WHEN THEN helps clarifying what is being tested and what is expected.

Replace should_pass_for_expected_equal_to_actual_in_certain_range1 and should_pass_for_expected_equal_to_actual_in_certain_range2 by a parameterized test as I already suggested

this should look like (with an additional parameter for precision)
```
  @ParameterizedTest
  @CsvSource({ "1.0f, 1.0f", "0.0f, 0.0f" })
  void should_pass_using_primitive_comparison(Float actual, Float expected) {
    then(actual).isEqualTo(expected);
  }
```

YeeTone replied:
I have fixed this problem.

joel-costigliola commented:
@YeeTone I have reviewed the PR, still quite a bit work to do, thanks!

YeeTone commented:
> @YeeTone I have reviewed the PR, still quite a bit work to do, thanks!

Could you tell me how to improve the PR? I sincerely hope my PR can be merged.

joel-costigliola commented:
> Could you tell me how to improve the PR? I sincerely hope my PR can be merged.

I already have commented your PR, read each comment and address them or ask me for clarification if it's not clear enough.

YeeTone commented:
> > Could you tell me how to improve the PR? I sincerely hope my PR can be merged.
> 
> I already have commented your PR, read each comment and address them or ask me for clarification if it's not clear enough. The PR is not ready to be merged until all the comments are addressed.

Thank you very much!

YeeTone commented:
Could you trigger the test for my PR? Thanks a lot! @scordio

YeeTone commented:
Sorry, I did not pull the latest version and became 'out of date', thus the checks are failed. Could you do me a favour to rerun the check? I'm sorry for the inconvenience. @scordio

YeeTone commented:
I found that the tests in org.assertj.core.error.ShouldHaveSameFileSystemAs_create_Test and org.assertj.core.error.ShouldHaveFileSystem_create_Test are all failed. And I found this is caused by back-slash(\) and slash(/), so I also fixed them. Please help me trigger the test again. Thx. @scordio

YeeTone commented:
I found that the tests in org.assertj.core.error.ShouldHaveSameFileSystemAs_create_Test and org.assertj.core.error.ShouldHaveFileSystem_create_Test are all failed. And I found this is caused by back-slash(\) and slash(/)。This is not easy since it depends on the operating system and I also fixed them. Please help me trigger the test again. Thx. @scordio

YeeTone commented:
I have passed most of the checks except the pitest. Could you do me a favour to have a look on my PR? Thank you very much! @joel-costigliola

YeeTone commented:
In the latest commit, I have already refactored the previous implementation using `BigDeciaml` and `String.valueOf` to avoid float/double precision bugs. Moreover, I fixed the bugs in `org.assertj.core.error.ShouldHaveSameFileSystemAs_create_Test` and `org.assertj.core.error.ShouldHaveFileSystem_create_Test`. Could you do me a favour to have a look at my latest PR? I have already passed the checks except the pitest(it seems to be a bug in pitest). @joel-costigliola

joel-costigliola commented on src/main/java/org/assertj/core/api/ComparatorFactory.java:
```diff
@@ -12,21 +12,58 @@
+      return ((Double)number).isNaN() || ((Double)number).isInfinite();
+    }else {
+      return false;
+    }
```
@YeeTone please address this comment

joel-costigliola replied:
@YeeTone please address this comment and the following I already made in a previous review: you don't need `Objects.requireNonNull(number)` since instanceof returns false for null

YeeTone replied:
Solved!

joel-costigliola commented on src/main/java/org/assertj/core/api/ComparatorFactory.java:
```diff
@@ -12,21 +12,80 @@
+   * @param number Object of Float or Double class
+   * @param <T> type of expected and precision, which should be the subclass of java.lang.Number and java.lang.Comparable
+   * @return whether the object is Nan or Infinity
+   * @throws java.lang.NullPointerException if number is null
```
that's not what this method do

YeeTone replied:
> that's not what this method do

I do not understand. This is the javadoc of `isNanOrInfinity` method. Here are the overview of this method:

```java
/**
   * Judge whether the input number is Nan or Infinity
   * @param number Object of Float or Double class
   * @param <T> type of expected and precision, which should be the subclass of java.lang.Number and java.lang.Comparable
   * @return whether the object is Nan or Infinity
   * @throws java.lang.NullPointerException if number is null
   */
  public static <T extends Number & Comparable<T>> boolean isNanOrInfinity(T number){
    // avoid null argument
    if(number instanceof Float){
      return ((Float) number).isNaN() || ((Float) number).isInfinite();
    }else if(Objects.requireNonNull(number) instanceof Double){
      return ((Double)number).isNaN() || ((Double)number).isInfinite();
    }

    return false;

  }
```
Could you explain why this is not what this method do?

joel-costigliola replied:
you are correct, it does throw a NPE when number is null because of `Objects.requireNonNull(number)` but I asked you not to call `Objects.requireNonNull(number)` as `instanceof` checks for null and returning `false` is correct if number is null

joel-costigliola commented on src/test/java/org/assertj/core/error/ShouldHaveFileSystem_create_Test.java:
```diff
@@ -12,18 +12,19 @@
+import static org.assertj.core.api.BDDAssertions.then;
+import static org.assertj.core.error.ShouldHaveFileSystem.shouldHaveFileSystem;
+import static org.mockito.BDDMockito.given;
+import static org.mockito.Mockito.mock;
```
revert all these changes, this has nothing to do with your PR

joel-costigliola commented on src/test/java/org/assertj/core/error/ShouldHaveSameFileSystemAs_create_Test.java:
```diff
@@ -12,15 +12,16 @@
-import org.junit.jupiter.api.Test;
+import static java.lang.String.format;
+import static org.assertj.core.api.BDDAssertions.then;
+import static org.assertj.core.error.ShouldHaveSameFileSystemAs.shouldHaveSameFileSystemAs;
```
revert all these changes, this has nothing to do with your PR

YeeTone replied:
These changes are to fix the bugs in these tests(Some checks are failed due to this bug). And this just change the import sequence. It should not have any other impact?😂

joel-costigliola replied:
this PR is about float/double not `ShouldHaveSameFileSystemAs`, don't mix things up.

and TBH it's not funny for me to spend so much time reviewing this PR.

joel-costigliola commented on src/test/java/org/assertj/core/internal/floats/Floats_assertIsCloseTo_Test.java:
```diff
@@ -75,6 +79,23 @@ void should_pass_if_actual_and_expected_are_NaN() {
+      byLessThan(0.25f)));
+    assertThatExceptionOfType(AssertionError.class).isThrownBy(() -> floats.assertIsCloseTo(someInfo(), 0.375f, 0.125f,
+      within(0.2499999f)));
+  }
```
replace `assertThatExceptionOfType(AssertionError.class).isThrownBy` by `expectAssertionError` as indicated in https://github.com/assertj/assertj-core/blob/main/CONTRIBUTING.md and do that everywhere you used it.

YeeTone replied:
Oh. I forgot this! I will fix this soon! Thx!

joel-costigliola commented on src/test/java/org/assertj/core/api/ComparatorFactory_doubleComparatorWithPrecision.java:
```diff
@@ -0,0 +1,87 @@
+    assertThat(comparator0_1d.compare(expected, actual)).isEqualTo(0);
+    assertThat(comparator0_1d.compare(actual, expected)).isEqualTo(0);
+    assertThat(comparator0_1d.compare(actual + precision / deltaParameter, expected)).isEqualTo(1);
+    assertThat(comparator0_1d.compare(expected, actual + precision / deltaParameter)).isEqualTo(-1);
```
the point of parameterized tests is to add parameters value, if you have only one it is basically useless.

the parameters should be: actual, expected and precision:
```java
  @ParameterizedTest
  @CsvSource({"0.111, 0.01, 100",
              "0.01, 0.111, 100",
              // add as many combination you can think of to cover edge cases
              "0.01, 0.111, 100"})
  void should_evaluate_actual_and_expected_as_equals(Double actual, Double expected, Double precision) {
    // GIVEN
    Comparator<Double> comparator = INSTANCE.doubleComparatorWithPrecision(precision);
    // WHEN/THEN
    then(comparator.compare(actual, expected)).isZero();
}
```
let me know if you don't understand.

YeeTone replied:
This is a good suggestion! I have understood it, and will work on it.

joel-costigliola commented:
@YeeTone I already have looked at your PR, I have made quite a few comments you haven't addressed, have you seen them?
What you need to do is address them or ask for clarifications if you don't understand them.

YeeTone commented:
> @YeeTone I already have looked at your PR, I have made quite a few comments you haven't addressed, have you seen them? What you need to do is address them or ask for clarifications if you don't understand them.

I also fixed the bugs in `org.assertj.core.error.ShouldHaveSameFileSystemAs_create_Test` and `org.assertj.core.error.ShouldHaveFileSystem_create_Test`. Is there any suggestion to the fixing on these 2 tests?

YeeTone commented:
> @YeeTone I already have looked at your PR, I have made quite a few comments you haven't addressed, have you seen them? What you need to do is address them or ask for clarifications if you don't understand them.

I also fixed the bugs in `org.assertj.core.error.ShouldHaveSameFileSystemAs_create_Test` and `org.assertj.core.error.ShouldHaveFileSystem_create_Test`. Is there any suggestion on the fixing of these 2 tests? 
@joel-costigliola

YeeTone commented:
For the NaN and Infinity cases, my idea is:
```
compare(expected, actual, precision)
if precision is nan or infinity:
     throw IllegalArgumentException(...)//Some information
if (expected is nan or infinity) or (actual is nan or infinity):
     return Double.compare(expected, actual)
return compareAsBigDecimalWithPrecision(expected, actual, precision)
```
In this case, the compare method can handle the case and get the expected output, for instance:
+Nan, +Nan, 0 -> 0
+Nan, -Nan, 0 -> 1
+Infinity, -Infinity -> 1
etc.

Could you give me some advice on my comparing idea? I am not quite sure. Thank you very much. @joel-costigliola

joel-costigliola commented:
For Nan/Infinity let's follow the JDK behavior with `Double.compare` as you suggested.
You can comment directly on the PR instead of here, it's easier to follow for reviewers.

YeeTone commented:
Could you do me a favor to run the test for PR? @scordio

YeeTone commented:
I found the tests failed due to bugs in the test files. 
In the latest commit, I did:
(1) revert the changes in `ShouldHaveFileSystem_create_Test` and `ShouldHaveSameFileSystemAs_create_Test`
(2) fix bugs in the testcases
(3) fix the code with `Objects.requireNonNull`
(4) let the comparing between NaN or Infinity follow the JDK behaviour

Sorry for my inconvenience. But sincerely hope the test can be run again.  @scordio

YeeTone commented:
I found the tests failed due to bugs in the test files. 
In the latest commit, I did:
(1) revert the changes in `ShouldHaveFileSystem_create_Test` and `ShouldHaveSameFileSystemAs_create_Test`
(2) fix bugs in the testcases
(3) fix the code with `Objects.requireNonNull`
(4) let the comparing between NaN or Infinity follow the JDK behaviour
(5) refactor the relevant method name

Sorry for my inconvenience. But sincerely hope the test can be run again.  @scordio

YeeTone commented:
Could you trigger the test for the PR? Thanks a lot! @scordio

YeeTone commented:
In the latest commit, I did:
(1) revert the changes in `ShouldHaveFileSystem_create_Test` and `ShouldHaveSameFileSystemAs_create_Test`
(2) fix bugs in the testcases
(3) fix the format problem from IntelliJ IDEA
(4) fix the code with `Objects.requireNonNull`
(5) let the comparing between NaN or Infinity follow the JDK behaviour
(6) refactor the relevant method name

And currently, I have passed most checkings except pitest. Could you do me a favor to review my latest PR? Sorry for the inconvenience. @joel-costigliola

YeeTone commented:
Thanks for all the previous suggestions! Could you have a look on my latest PR? I have followed all your advice, and I sincerely hope you could give more comments on the PR.😁 @joel-costigliola

YeeTone commented:
If the comments could be given directly on the PR, then it will be more efficient for me to take the suggestion.

joel-costigliola commented on src/test/java/org/assertj/core/api/ComparatorFactory_doubleComparatorWithPrecision.java:
```diff
@@ -0,0 +1,97 @@
+    assertThat(comparator0_1d.compare(max, min)).isEqualTo(0);
+    assertThat(comparator0_1d.compare(min, max)).isEqualTo(0);
+    assertThat(comparator0_1d.compare(max + smallDifference, min)).isEqualTo(1);
+    assertThat(comparator0_1d.compare(min - smallDifference, max)).isEqualTo(-1);
```
as commented previously in https://github.com/assertj/assertj-core/pull/2535#discussion_r847202709 this should have only one assertion. I basically wrote the code for that tests, you only have to populate the CsvSource.

joel-costigliola commented on src/test/java/org/assertj/core/api/ComparatorFactory_doubleComparatorWithPrecision.java:
```diff
@@ -0,0 +1,97 @@
+import static org.assertj.core.error.ShouldBeEqualWithinOffset.shouldBeEqual;
+import static org.assertj.core.util.AssertionsUtil.expectAssertionError;
+
+public class ComparatorFactory_doubleComparatorWithPrecision {
```
test classes don't need to be public, they need to end with `_Test`  rename the class and the file

joel-costigliola commented on src/test/java/org/assertj/core/internal/doubles/Doubles_assertIsCloseTo_Test.java:
```diff
@@ -39,13 +45,24 @@ class Doubles_assertIsCloseTo_Test extends DoublesBaseTest {
+    doubles.assertIsCloseTo(someInfo(), 0.375d, 0.125, within(0.25));
+    doubles.assertIsCloseTo(someInfo(), 0.125d, 0.375, byLessThan(0.2500001));
+    doubles.assertIsCloseTo(someInfo(), 1.1d, 1.0d, within(0.1d));
+    doubles.assertIsCloseTo(someInfo(), 1.0d, 1.1d, byLessThan(0.1000001d));
```
use parameterized test and have one for `within` and another for `byLessThan`

joel-costigliola reviewed (changes requested):
@YeeTone there were some unaddressed comments from my previous review.

Before asking me to review again, you should address all the comments or ask for clarifications, I have spent a lot of time already reviewing this PR.

YeeTone commented:
Thanks a lot for your patience! I will address the comment soon. @joel-costigliola

YeeTone commented:
Could I make `isNanOrInfinity` method in the ComparatorFactory public? Because some classes such as RealNumbers may utilize it to avoid code duplication.

joel-costigliola commented:
I would rather have it in `RealNumbers`

YeeTone commented:
> I would rather have it in `RealNumbers`

OK! Thx a lot!

YeeTone commented:
Could I define the following in the RealNumbers class to implement the function comparing Nan or Infinity?
```java
public boolean isNanOrInfinite(NUMBER value){
    return isNaN(value) || isInfinite(value);
  }
```
 while `isNaN` and `isInfinite` are abstract methods so that they are implemented by the subclass Floats or Doubles?

joel-costigliola commented:
Yes, that was polymorphism is about.

YeeTone commented:
Thanks a lot for your patience! But I am a little busy recently, and could you give me a little more time for the final PR? Sorry for inconvenience.

YeeTone commented on src/main/java/org/assertj/core/api/ComparatorFactory.java:
```diff
@@ -12,21 +12,58 @@
+    // java.math.BigDecimal cannot handle NAN or Infinity cases.
+    if(isNanOrInfinity(expected) || isNanOrInfinity(actual)){
+      return false;
+    }
```
Fixed!

YeeTone commented on src/main/java/org/assertj/core/api/ComparatorFactory.java:
```diff
@@ -12,21 +12,58 @@
+    }
+  }
+
+  public static <T extends Number & Comparable<T>> BigDecimal preciseBigDecimal(T number) {
```
In the latest design, there are only 2 public methods, `doubleComparatorWithPrecision` and `floatComparatorWithPrecision`

YeeTone reviewed (commented):
I make sure that I have read and resolved all conversation of @joel-costigliola.

YeeTone commented:
Could you do me a favour to run the test for the PR? Thx! @scordio

YeeTone commented:
I also reviewed all my commit and confirmed that I had addressed all the comments and done the modification. Could you review it after the testes of github PR have been passed? Thanks for you patience. @joel-costigliola

joel-costigliola commented on src/test/java/org/assertj/core/api/ComparatorFactory_doubleComparatorWithPrecision_Test.java:
```diff
@@ -0,0 +1,166 @@
+              "1.2463, 1.2464, 0.0001, 1e-8"})
+  void should_pass_for_expected_equal_to_actual_in_certain_range_max_min(Double expected, Double actual, Double precision, Double smallDifference) {
+    Comparator<Double> comparator = INSTANCE.doubleComparatorWithPrecision(precision);
+    Double max = Math.max(expected, actual), min = Math.min(expected, actual);
```
why do you need `max` and `min`?

this test should be as simple as: 
```
  // GIVEN
  Comparator<Double> comparator = INSTANCE.doubleComparatorWithPrecision(precision);
  // THEN
  then(comparator.compare(actual, expected)).isZero();
```
why make things more complicated than they should be?

joel-costigliola replied:
this comment applies to the other test too

YeeTone replied:
Oh, I tried to test whether the order of bigger and smaller value will affect the result. I will simplify the assertion!

joel-costigliola replied:
just use more values in the CsvSource

YeeTone replied:
Fixed!

joel-costigliola reviewed (changes requested):
still more work to do @YeeTone

YeeTone commented on src/main/java/org/assertj/core/api/ComparatorFactory.java:
```diff
@@ -12,21 +12,87 @@
+   * @return whether the object is Nan or Infinity
+   * @throws java.lang.NullPointerException if number is null
+   */
+  private static <T extends Number & Comparable<T>> boolean isNanOrInfinity(T number) {
```
I thought that since T is a generic type, the judgement depends on the type of T. If I remove this method, can I modify the public compare method like this?
```java
    public int compare(Double o1, Double o2) {
        if(isNanOrInfinity(precision)){
          throw new IllegalArgumentException("Precision should not be Nan or Infinity!");
        }

        if(Doubles.instance().isNanOrInfinite(o1) || Doubles.instance().isNanOrInfinite(o2)){
          return Double.compare(o1, o2);
        }
        if (compareAsBigDecimalWithPrecision(o1, o2, precision)) {
          return 0;
        }
        return Double.compare(o1, o2);
      }

```

YeeTone replied:
Or is it a better idea to define a field so that the duplicated call to `Doubles.instance()` can be avoided?

YeeTone replied:
I have removed the `isNanOrInfinite` method in the class `ComparatorFactory` and replace it as the call of `Doubles.instance().isNanOrInfinite`.

YeeTone commented:
In the [latest one commit](https://github.com/assertj/assertj-core/pull/2535/commits/e83c9916c937a98bbfeedb50bd6b0778c7419304), I have refactored the code based on the suggestions! Could you give me more code review? @joel-costigliola

joel-costigliola commented on src/test/java/org/assertj/core/api/ComparatorFactory_doubleComparatorWithPrecision_Test.java:
```diff
@@ -0,0 +1,162 @@
+    "1.2464, 1.2463, 9e-5"})
+  void should_fail_for_expected_equal_to_actual_not_in_certain_range_max_min(Double expected, Double actual, Double precision) {
+    Comparator<Double> comparator = INSTANCE.doubleComparatorWithPrecision(precision);
+    Double max = Math.max(expected, actual), min = Math.min(expected, actual);
```
There are still a bunch of useless min max variables, refactor ALL the tests to get rid of them

joel-costigliola commented on src/test/java/org/assertj/core/api/ComparatorFactory_doubleComparatorWithPrecision_Test.java:
```diff
@@ -0,0 +1,162 @@
+    "0.7654320, 0.7654321, 0.0000001",
+    "1.2463, 1.2464, 0.0001",
+    "1.2464, 1.2463, 0.0001"})
+  void should_pass_for_expected_equal_to_actual_in_certain_range_max_min(Double expected, Double actual, Double precision) {
```
the test name should not contain min max (same applies for other tests)

YeeTone replied:
Sorry, I did not see it. I have modified the test name.

joel-costigliola commented on src/main/java/org/assertj/core/internal/RealNumbers.java:
```diff
@@ -93,5 +100,13 @@ public void assertIsNotInfinite(AssertionInfo info, NUMBER actual) {
+
+  protected abstract boolean isNaN(NUMBER value);
+
+  protected abstract boolean isNotNaN(NUMBER value);
```
are you using it? if not remove it

YeeTone replied:
I wrote it because I saw the previous method definition such as ’isFinite‘ and `isNotFinite`. I have removed it.

joel-costigliola commented on src/main/java/org/assertj/core/internal/RealNumbers.java:
```diff
@@ -93,5 +100,13 @@ public void assertIsNotInfinite(AssertionInfo info, NUMBER actual) {
 
+  public boolean isNanOrInfinite(NUMBER value){
+    return isNaN(value) || isInfinite(value);
+  }
```
write a parameterized test for this method

YeeTone replied:
I would like to provide a javadoc for it. Is it a good idea?

YeeTone commented:
Thanks for your review to the PR! I make sure that I have viewed all the recent commits and made modification based on the suggestions. I sincerely hope you could give more advice on the PR. @joel-costigliola

YeeTone commented:
Thanks for your review to the PR! I make sure that I have viewed all the recent commits and made modification based on the suggestions. I sincerely hope you could give more advice on the PR. @joel-costigliola

YeeTone commented:
Thanks for your review to the PR! I make sure that I have viewed all the recent commits and made modification based on the suggestions. I sincerely hope you could give more advice on the PR. @joel-costigliola

YeeTone commented:
Currently, I have reviewed **ALL** your comments and modified based on the suggestions. And the running check also has been passed. Could you do me a favour to give more advice? @joel-costigliola

YeeTone commented:
Currently, I have reviewed **ALL** your comments and done modification based on the suggestions. And the running checks also have been passed. Could you do me a favour to give more advice? @joel-costigliola

YeeTone commented:
Currently, I have reviewed **ALL** your comments and done modification based on the suggestions. And the running checks also have been passed. Could you do me a favour to give more advice? Thx! @joel-costigliola

YeeTone commented:
Currently, I have reviewed **ALL** your comments and done modification based on the suggestions. And **ALL** the running checks also have been passed. Could you do me a favour to give more advice? Please let me know if there is some work to do for this PR! @joel-costigliola

joel-costigliola commented:
I'm currently integrating this PR (and refactor it), thanks for it @YeeTone.

YeeTone commented:
> I'm currently integrating this PR (and refactor it), thanks for it @YeeTone.

Thanks a lot for your patience and review!

joel-costigliola commented:
Integrated thanks @YeeTone!
